pgrx/src/callconv.rs

```diff
@@ -283,3 +284,170 @@ where
+    std::marker::PhantomData<&'fcx mut FcInfoData>,
+);
+
+// when talking about this, there's the lifetime for setreturningfunction, and then there's the current context's lifetime.
+// Potentially <'srf, 'curr, 'ret: 'curr + 'srf> -> <'ret> but don't start with that.
+// at first <'curr> or <'fcx>
+// It's a heap-allocated stack frame for your function.
+// 'fcx would apply to the arguments into this whole thing.
+// PgHeapTuple is super not ideal and this enables a replacement of that.
+// ArgAbi and unbox from fcinfo index
+// Just this and the accessors, something that goes from raw_args(&'a self) -> &'fcx [NullableDatum]? &'a [NullableDatum]?
+impl<'fcx> FcInfo<'fcx> {
+    /// Constructor, used to wrap a raw FunctionCallInfo provided by Postgres.
+    ///
+    /// # Safety
+    ///
+    /// This function is unsafe as we cannot ensure the `fcinfo` argument is a valid
+    /// [`pg_sys::FunctionCallInfo`] pointer.  This is your responsibility.
+    pub unsafe fn assume_valid(fcinfo: pg_sys::FunctionCallInfo) -> FcInfo<'fcx> {
+        let _nullptr_check =
+            std::ptr::NonNull::new(fcinfo).expect("fcinfo pointer must be non-null");
+        Self(fcinfo, std::marker::PhantomData)
+    }
+    /// Retrieve the arguments to this function call as a slice of [`pgrx_pg_sys::NullableDatum`]
+    #[inline]
+    pub fn raw_args<'a>(&'a self) -> &'fcx [pgrx_pg_sys::NullableDatum]
+    where
+        'a: 'fcx,
+    {
+        // Null pointer check already performed on immutable pointer
+        // at construction time.
+        unsafe {
+            let arg_len = (*self.0).nargs;
+            let args_ptr: *const pg_sys::NullableDatum = std::ptr::addr_of!((*self.0).args).cast();
+            // A valid FcInfoWrapper constructed from a valid FuntionCallInfo should always have
+            // at least nargs elements of NullableDatum.
+            std::slice::from_raw_parts(args_ptr, arg_len as _)
+        }
+    }
+
+    /// Modifies the contained `fcinfo` struct to flag its return value as null.
+    ///
+    /// Returns a null-pointer Datum for use with the calling function's return value.
+    ///
+    /// Safety: If this flag is set, regardless of what your function actually returns,
+    /// Postgress will presume it is null and discard it. This means that, if you call this
+    /// method and then return a value anyway, *your extension will leak memory.* Please
+    /// ensure this method is only called for functions which, very definitely, returns null.
+    ///
+    /// # Examples
+    ///
+    /// ```rust,no_run
+    /// use pgrx::callconv::FcInfo;
+    /// use pgrx::prelude::*;
+    /// fn foo<'a>(mut fcinfo: FcInfo<'a>) -> pg_sys::Datum {
+    ///     return unsafe { fcinfo.set_return_null() };
+    /// }
+    /// ```
+    #[inline]
+    pub unsafe fn set_return_null(&mut self) -> pg_sys::Datum {
+        let fcinfo = unsafe { self.0.as_mut() }.unwrap();
+        fcinfo.isnull = true;
+        pg_sys::Datum::from(0)
+    }
```

Comment:
Should return a `pgrx::datum::Datum<'a>`

Reply:
Resolved